## /bin/sh -e simulate dpatch for prepend_chlog.py
## Feature
## ukcid.dpatch by Tzafrir Cohen <tzafrir.cohen@xorcom.com>
## DP: A backport of playdtmf. See http://bugs.digium.com/view.php?id=6682
## DP: patch: svn diff -r 10646:13356 http://svn.digium.com/svn/asterisk/team/oej/test-this-branch/apps/app_senddtmf.c
## DP: From: http://www.lusyn.com/asterisk/patches.html

Index: asterisk-1.2.27/apps/app_senddtmf.c
===================================================================
--- asterisk-1.2.27.orig/apps/app_senddtmf.c	2008-03-19 11:32:21.000000000 +0100
+++ asterisk-1.2.27/apps/app_senddtmf.c	2008-03-19 11:33:29.000000000 +0100
@@ -41,6 +41,7 @@
 #include "asterisk/options.h"
 #include "asterisk/utils.h"
 #include "asterisk/app.h"
+#include "asterisk/manager.h"
 
 static char *tdesc = "Send DTMF digits Application";
 
@@ -95,11 +96,39 @@
 	return res;
 }
 
+static char mandescr_playdtmf[] =
+	"Description: Plays a DTMF digit on the specified channel.\n"
+	"Variables: (all are required)\n"
+	"Channel: Channel name to send digit to\n"
+	"Digit: The dtmf digit to play\n";
+
+static int manager_play_dtmf(struct mansession *s, struct message *m)
+{
+	char *channel, *digit;
+
+	channel = astman_get_header(m, "Channel");
+	digit = astman_get_header(m, "Digit");
+	struct ast_channel *chan = ast_get_channel_by_name_locked(channel);
+	if (chan == NULL) {
+		astman_send_error(s, m, "No such channel");
+		return 0;
+	}
+	if (digit == NULL) {
+		astman_send_error(s, m, "No digit specified");
+		return 0;
+	}
+	ast_senddigit(chan, *digit);
+	ast_mutex_unlock(&chan->lock);
+	astman_send_ack(s, m, "DTMF successfully sent");
+	return 0;
+}
+
 int unload_module(void)
 {
 	int res;
 
 	res = ast_unregister_application(app);
+	res |= ast_manager_unregister("playDTMF");
 
 	STANDARD_HANGUP_LOCALUSERS;
 
@@ -108,6 +137,7 @@
 
 int load_module(void)
 {
+	ast_manager_register2( "playDTMF", EVENT_FLAG_AGENT, manager_play_dtmf, "Play DTMF signal on a specific channel.", mandescr_playdtmf);
 	return ast_register_application(app, senddtmf_exec, synopsis, descrip);
 }
 
